tactics/tactics.ml

```diff
@@ -889,23 +940,33 @@ let reduce redexp cl =
   in
   Proofview.Trace.name_tactic trace begin
   Proofview.Goal.enter begin fun gl ->
+  let env = Proofview.Goal.env gl in
   let hyps = concrete_clause_of (fun () -> Tacmach.New.pf_ids_of_hyps gl) cl in
   let nbcl = (if cl.concl_occs = NoOccurrences then 0 else 1) + List.length hyps in
   let check = match redexp with Fold _ | Pattern _ -> true | _ -> false in
-  let reorder = match redexp with Fold _ | Pattern _ -> true | _ -> false in
+  let reorder = match redexp with
+  | Fold _ | Pattern _ -> AnyHypConv
+  | Simpl (flags, _) | Cbv flags | Cbn flags | Lazy flags ->
+    if is_local_flag env flags then LocalHypConv else StableHypConv
+  | Unfold flags ->
+    if is_local_unfold env flags then LocalHypConv else StableHypConv
+  | Red _ | Hnf | CbvVm _ | CbvNative _ -> StableHypConv
+  | ExtraRedExpr _ -> StableHypConv (* Should we be that lenient ?*)
```

Comment:
What does this mean?

Reply:
That this is preserving the previous behaviour, but that external reductions need not fulfill this extraneous properties in any way, leading to potentially incorrect results. I don't know if we rely too much on the extensibility of reduction functions in practice though.

Reply:
I see. Please either fix it here or open an issue, it seems rather serious (and hard to debug for clients).

Reply:
I'll open an issue, otherwise fixing this trivially might trigger slowdowns for legit users.